### DISCUSSION TITLE
How to build transmission-cli for windows on Cygwin

### DISCUSSION THREAD

**Discussion by Limitex** [transmission/transmission#2108] [General]
How to build transmission-cli for windows on Cygwin.
I have tried the following.

0. Enable file system case sensitivity in the registry. [Described later]
1. Install Cygwin on Windows.
2. Select the Cygwin package [Described later]
3. Launch Cygwin64 Terminal
4. Move to any directory. [cd ~~~]
5. Download file [Described later]
6. expansion [Described later]
7. Move the working directory [Described later]
8. create a build directory and move [Described later] 
9. Run cmake [Described later] 
10. Run make [make] <<<< ERROR (make: *** No targets specified and no makefile found.  Stop.)

Apparently it can't find the makefile, what should I do?

* case sensitivity

```registry:ntfs_casesensitive.reg
Windows Registry Editor Version 5.00

[HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Session Manager\kernel]
"obcaseinsensitive"=dword:00000000
```

* Installed Cygwin package

```
wget
gcc-core
gcc-g++
make
cmake
texinfo
libmpc-devel
bison, flex
libelf-devel
bc
curl
libcurl-devel
diffutils
```

* Executed commands

```
$ wget https://github.com/transmission/transmission/releases/download/3.00/transmission-3.00.tar.xz
$ tar -Jxvf transmission-3.00.tar.xz
$ cd transmission-3.00
$ mkdir build
$ cd build
$ cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo ..
$ make
```

and, Is it possible to make the executable file a single file? [transmission-cli.exe only]

**Comment by depler** [comment]
You can try this instead: https://github.com/depler/transmission-vs-beta. Just replace `daemon.cc` file with `cli.cc` and build standalone binary (Visual Studio 2019 required).

**Comment by Limitex** [comment]
thank you.
Is this a port of the Transmission source code to the Visual studio Project as it is?
Also, please tell me the original method.

**Comment by depler** [comment]
> Is this a port of the Transmission source code to the Visual studio Project as it is?

Yes, it is also contains some small fixes and updated third-party components. You can replace all source code files with original transmission files if you warn about it.

> Also, please tell me the original method.

I don't use original method with cygwin due to build complexity, external dependencies and outdated FAQ. So I cannot help you in that.

If you just need ready to use binaries, take a look at stable 3.00 version: https://github.com/depler/transmission-vs/releases. It does not contains `cli`, but instead it has modern web UI transmissionic, which allow to control almost everything.

**Comment by Limitex** [comment]
I see that Transmission 3.00 does not include cli.
You have multiple Transmissions in your repositories, which one is the best?

how can I build a Transmission-cli.exe for Windows with this VS project?
Also, Is it possible to generate a single file?
example [transmission-cli.exe only]
No cli.cc found.
Sorry for so many questions.

**Comment by depler** [comment]
If you plan to use transmission on everyday basis - take version 3.00. The other one is beta version with latest code changes, it is still under construction. 

How to build cli:
1. Take desired version of cli.c or cli.cc. For version 3.00 it would this one: https://github.com/transmission/transmission/blob/3.00/cli/cli.c
2. Open transmission.sln in visual studio, select Release configuration and build it - this is how you check that everything is working ok
3. Find file daemon.c and delete it from project
4. Add file cli.c to project and build it again - now you have single file transmission.exe in binary folder, which is actually cli (you can rename it)

I didn't add cli.c into solution, because this leads to creating one more project with huge amount of settings.

**Comment by Limitex** [comment]
Thanks for explaining in detail!
https://github.com/depler/transmission-vs
Is this the source code for the release of this?

**Comment by depler** [comment]
This is stable version 3.00
